src/Database/LSMTree/Monoidal.hs

```diff
@@ -325,8 +334,75 @@ duplicate = undefined
 -- NOTE: merging table handles creates a new table handle, but does not close
 -- the table handles that were used as inputs.
 merge ::
-     (IOLike m, SomeUpdateConstraint v)
+     (IOLike m, ResolveValue v)
   => TableHandle m k v
   -> TableHandle m k v
   -> m (TableHandle m k v)
 merge = undefined
+
+{-------------------------------------------------------------------------------
+  Monoidal value resolution
```

Comment:
Ah, yes. That's also something I was thinking about. It feels like a class is the right API, especially since there are laws involved. But then we need to see how that fits in with the table config.

Reply:
Removing it from the config would be nice, because then we don't have to write the function to disk on a snapshot. A class makes less sense for an untyped API (if we decide to add that at some point)

Reply:
In the `Internal` module, we'd still need to take the `ResolveMupsert` as an argument (so `Monoidal` grabs it from the class and passes it explicitly), which adds some verbosity, but you're right, it should not be part of the table config type that gets serialised either way.